ddraw/IDirect3DDeviceX.cpp

```diff
@@ -338,14 +338,96 @@ HRESULT m_IDirect3DDeviceX::SetTransform(D3DTRANSFORMSTATETYPE dtstTransformStat
 			break;
 		}
 
-		HRESULT hr = (*d3d9Device)->SetTransform(dtstTransformStateType, lpD3DMatrix);
-
-		if (SUCCEEDED(hr))
+		D3DMATRIX view;
+		if (Config.DdrawConvertHomogeneousW)
 		{
 #ifdef ENABLE_DEBUGOVERLAY
+			// Set the original matrix
 			DOverlay.SetTransform(dtstTransformStateType, lpD3DMatrix);
 #endif
+
+			if (dtstTransformStateType == D3DTS_VIEW)
+			{
+				D3DVIEWPORT9 Viewport9;
+				if (SUCCEEDED((*d3d9Device)->GetViewport(&Viewport9)))
+				{
+					const float width = (float)Viewport9.Width;
+					const float height = (float)Viewport9.Height;
+
+					// Replace the matrix with one that handles D3DFVF_XYZRHW geometry
+					ZeroMemory(&view, sizeof(D3DMATRIX));
+					view._11 = 2.0f / width;
+					view._22 = -2.0f / height;
+					view._33 = 1.0f;
+					view._41 = -1.0f;  // translate X
+					view._42 = 1.0f;   // translate Y
+					view._44 = 1.0f;
+
+					// Backup pointer before overriding original matrix pointer
+					LPD3DMATRIX lpOriginalD3DMatrix = lpD3DMatrix;
```

Comment:
This intermediate variable can be avoided if the `lpD3DMatrix` overwrite is moved near the end of this statement block.
That way, here below code uses `lpOriginalD3DMatrix`, it could still use `lpD3DMatrix`.
The only other thing that would have to change, is the one occurrence were below code currently uses `lpD3DMatrix`, it should use `view` instead.

Reply:
Good idea.  Done in the latest check-in.